tests/proxy-server.ts

```diff
@@ -0,0 +1,217 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * This module creates two HTTP servers listening on the local machine:
+ * * The "proxy server" which is a standard HTTP proxy server that handles the
+ *   CONNECT HTTP verb, using the `http-proxy` dependency.
+ * * The "interceptor server" which actually handles the proxied requests.
+ *
+ * The proxy server proxies the client request to the interceptor server. (This
+ * two-server approach (proxy + interceptor) is mainly a result of accommodating
+ * the typical setup documented by the `http-proxy` dependency.)
+ *
+ * The use case for these servers is to test the standalone executable (CLI's
+ * standalone zip package) in a child process. Most of the CLI's automated tests
+ * currently test HTTP requests using `nock`, but `nock` can only mock/test the
+ * same process (Node's built-in `http` library). However, the CLI has support
+ * for proxy servers as a product feature, so the idea was to proxy the child
+ * process requests to the parent process, where the proxy / interceptor servers
+ * run. The interceptor server then forwards the request (mostly unchanged) with
+ * the expectation that `nock` will intercept the requests for testing (in the
+ * parent process) as usual.
+ *
+ * 1. A `mocha` test case calls `runCommand('push test-rpi')`, with `nock` setup
+ *    to intercept HTTP requests (in the same process that runs `mocha`).
+ * 2. The proxy and interceptor servers are started in the parent process (only
+ *    once: singleton) at free TCP port numbers randomly allocated by the OS.
+ * 3. A CLI child process gets spawned to run the command (`balena push test-rpi`)
+ *    with environment variables including BALENARC_PROXY (set to
+ *    'http://127.0.0.1:${proxyPort}'). (Additional env vars instruct the
+ *    child process to use HTTP instead of HTTPS for the balena API and builder.)
+ * 4. The child process sends the HTTP requests to the proxy server.
+ * 5. The proxy server forwards the request to the interceptor server.
+ * 6. The interceptor server simply re-issues the HTTP request (unchange), with
+ *    the expectation that `nock` will intercept it.
+ * 7. `nock` (running on the parent process, same process that runs `mocha`)
+ *    intercepts the HTTP request, test it and replies with a mocked response.
+ * 8. `nocks` response is returned to the interceptor server, which returns it
+ *    to the proxy server, which returns it to the child process, which continues
+ *    CLI command execution.
```

Comment:
Check these comments for an explanation of the standalone executable tests involving a proxy server.